File: LowLevelDesign/java/task-scheduler/src/com/taskscheduler/TaskSchedulerService.java
package com.taskscheduler;

import com.taskscheduler.observer.TaskExecutionObserver;
import com.taskscheduler.strategy.SchedulingStrategy;
import com.taskscheduler.task.Task;

import java.time.Duration;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.PriorityBlockingQueue;

public class TaskSchedulerService {
    private static TaskSchedulerService instance;
    private Thread[] workers;
    private volatile boolean isRunning=true;
    private final List<TaskExecutionObserver> observers = new ArrayList<>();
    private final PriorityBlockingQueue<ScheduledTask> taskQueue = new PriorityBlockingQueue<>();

    private TaskSchedulerService() {}

    public static TaskSchedulerService getInstance() {
        if(instance == null) {
            instance = new TaskSchedulerService();
        }
        return instance;
    }

    public void initialize(int workerCount) {
        workers = new Thread[workerCount];
        startWorkers();
    }

    private void startWorkers() {
        for(int i=0; i < workers.length; i++) {
            workers[i] = new Thread(this::runWorker, "WorkerThread-" + i);
            workers[i].setDaemon(true);
            workers[i].start();
        }
    }

    public void schedule(Task task, SchedulingStrategy strategy) {
        ScheduledTask scheduledTask = new ScheduledTask(task, strategy);
        taskQueue.put(scheduledTask);
    }

    private void runWorker() {
        while(isRunning) {
            try {
                ScheduledTask task = taskQueue.take(); // // waits here until an item is added
                LocalDateTime now = LocalDateTime.now();
                long waitTime = 0;
                if(task.getNextExecutionTime().isAfter(now)) {
                    waitTime = Duration.between(now, task.getNextExecutionTime()).toMillis();
                }
                // For 5 seconds from now, it is showing 4994
//                System.out.println(waitTime);

                if(waitTime > 0) {
                    // Wait for scheduled Time
                    Thread.sleep(waitTime);
                }
                execute(task);

            } catch(InterruptedException e) {
                Thread.currentThread().interrupt();
                break;
            }
        }
    }

    private void execute(ScheduledTask task) {
        observers.forEach(o -> o.onTaskStarted(task));
        try {
            task.getTask().execute();
            task.updateLastExecutionTime();
            observers.forEach(o -> o.onTaskCompleted(task));
        } catch (Exception e) {
            System.err.printf("Task %s failed with error: %s%n", task.getId(), e.getMessage());
            observers.forEach(o -> o.onTaskFailed(task, e));
        } finally {
            // Rescheduling Logic
            task.updateNextExecutionTime();
            if(task.hasMoreExecutions()) {
                taskQueue.put(task); // Re-queue for the next run
            } else {
                System.out.printf("Task %s has no more executions and will not be rescheduled.%n", task.getId());
            }
        }
    }

    public void addObserver(TaskExecutionObserver observer) {
        observers.add(observer);
    }

    public void shutDown() {
        isRunning = false;
        for(Thread worker: workers) {
            worker.interrupt();
        }
        System.out.println("Scheduler shut down");
    }

}

File: LowLevelDesign/task-scheduler-nodejs/src/TaskSchedulerService.js
import { ScheduledTask } from './ScheduledTask.js';
import { MinHeap } from '@datastructures-js/heap';

/**
 * TaskSchedulerService - Singleton service that manages task scheduling and execution
 * Uses a priority queue to manage scheduled tasks and async worker functions for execution
 * Note: These are async functions (not Node.js Worker threads) - they run concurrently on the same event loop
 */
export class TaskSchedulerService {
    static instance = null;

    constructor() {
        this.workers = [];
        this.isRunning = true;
        this.observers = [];
        // MinHeap with comparator function that uses ScheduledTask's compareTo method
        this.taskQueue = new MinHeap((a, b) => a.compareTo(b));
    }

    static getInstance() {
        if (TaskSchedulerService.instance === null) {
            TaskSchedulerService.instance = new TaskSchedulerService();
        }
        return TaskSchedulerService.instance;
    }

    initialize(workerCount) {
        this.workers = [];
        for (let i = 0; i < workerCount; i++) {
            this.startWorker(i);
        }
    }

    startWorker(workerId) {
        const worker = async () => {
            while (this.isRunning) {
                try {
                    const task = await this.takeTask();
                    if (!task) continue;

                    console.log(`[WORKER ${workerId}] Picked task ${task.getId()} from queue (scheduled for ${task.getNextExecutionTime()?.toISOString() || 'N/A'})`);

                    const now = new Date();
                    const waitTime = task.getNextExecutionTime() 
                        ? Math.max(0, task.getNextExecutionTime().getTime() - now.getTime())
                        : 0;

                    if (waitTime > 0) {
                        // Put task back and wait
                        console.log(`[WORKER ${workerId}] Task ${task.getId()} not ready yet, waiting ${waitTime}ms. Returning to queue.`);
                        this.offer(task);
                        await this.sleep(Math.min(waitTime, 100)); // Short sleep to avoid blocking
                        continue;
                    }

                    console.log(`[WORKER ${workerId}] Executing task ${task.getId()} now`);
                    await this.execute(task, workerId);
                } catch (error) {
                    if (error.message !== 'Scheduler stopped') {
                        console.error(`Worker ${workerId} error:`, error);
                    }
                    break;
                }
            }
        };
        this.workers.push(worker());
    }

    schedule(task, strategy) {
        const scheduledTask = new ScheduledTask(task, strategy);
        console.log(`[MAIN PROCESS] Scheduling task ${scheduledTask.getId()} for execution at ${scheduledTask.getNextExecutionTime()?.toISOString() || 'N/A'}`);
        this.offer(scheduledTask);
    }

    async execute(scheduledTask, workerId = null) {
        const workerInfo = workerId !== null ? `[WORKER ${workerId}]` : '[MAIN PROCESS]';
        console.log(`${workerInfo} Starting execution of task ${scheduledTask.getId()}`);
        
        // Notify observers
        this.observers.forEach(observer => observer.onTaskStarted(scheduledTask));

        try {
            await scheduledTask.getTask().execute();
            scheduledTask.updateLastExecutionTime();
            console.log(`${workerInfo} Task ${scheduledTask.getId()} execution completed successfully`);
            this.observers.forEach(observer => observer.onTaskCompleted(scheduledTask));
        } catch (error) {
            console.error(`${workerInfo} Task ${scheduledTask.getId()} failed with error: ${error.message}`);
            this.observers.forEach(observer => observer.onTaskFailed(scheduledTask, error));
        } finally {
            // Rescheduling Logic
            scheduledTask.updateNextExecutionTime();
            if (scheduledTask.hasMoreExecutions()) {
                console.log(`${workerInfo} Rescheduling task ${scheduledTask.getId()} for ${scheduledTask.getNextExecutionTime()?.toISOString() || 'N/A'}`);
                this.offer(scheduledTask); // Re-queue for the next run
            } else {
                console.log(`${workerInfo} Task ${scheduledTask.getId()} has no more executions and will not be rescheduled.`);
            }
        }
    }

    addObserver(observer) {
        this.observers.push(observer);
    }

    shutDown() {
        this.isRunning = false;
        console.log("Scheduler shut down");
    }

    // Priority Queue Operations (Min Heap using @datastructures-js/heap)
    offer(task) {
        this.taskQueue.insert(task);
    }

    async takeTask() {
        while (this.isRunning && this.taskQueue.size() === 0) {
            await this.sleep(10); // Wait for tasks
        }

        if (!this.isRunning && this.taskQueue.size() === 0) {
            throw new Error('Scheduler stopped');
        }

        if (this.taskQueue.size() === 0) return null;

        // Extract and return the minimum (highest priority) task
        return this.taskQueue.extractRoot();
    }

    sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
}
